cassandra-analytics-integration-tests/src/test/java/org/apache/cassandra/analytics/SharedClusterSparkIntegrationTestBase.java

```diff
@@ -171,28 +173,35 @@ public void checkSmallDataFrameEquality(Dataset<Row> expected, Dataset<Row> actu
         }
     }
 
-    public void validateWritesWithDriverResultSet(List<Row> sourceData, ResultSet queriedData,
-                                                  Function<com.datastax.driver.core.Row, String> rowFormatter)
+    public void validateWritesWithDriverResultSet(List<Row> sparkData, ResultSet driverData,
+                                                  Function<com.datastax.driver.core.Row, String> driverRowFormatter)
     {
-        Set<String> actualEntries = new HashSet<>();
-        queriedData.forEach(row -> actualEntries.add(rowFormatter.apply(row)));
+        Set<String> driverEntries = new HashSet<>();
+        driverData.forEach(row -> driverEntries.add(driverRowFormatter
+                .apply(row)
+                // Driver Codec writes "NULL" for null value. Spark DF writes "null".
+                .replace("NULL", "null")
+                // driver writes lists as [] and sets as {},
+                // whereas spark entries have the same type WrappedArray for both lists and sets
+                .replace('[', '{')
+                .replace(']', '}')));
```

Comment:
I think this logic should be encapsulated in the function itself, and not here.

Reply:
Which function do you mean? We repeat the same logic for all driver entries formatters for UDTs (and also Tuples which is coming up today). So keeping it in one place instead of repeating the same in each formatter.

Reply:
Moved out collections specific formatting to corresponding formatter, to avoid other tests getting impacted by chance